### PR TITLE
Add support for hot reload commands

### LINKED ISSUE #3548: Implement `hotReloadCapable`  flags from devfile v2 spec
https://github.com/devfile/kubernetes-api/issues/64

Let's implement support for `restartOnChange` field as defined in the above issue

/triage needs-information

### LINKED ISSUE #3784: Devfile run command not started if `odo push` is run after `odo push --debug`
/kind bug

## What versions of software are you using?

**Operating System:**

**Output of `odo version`:**

## How did you run odo exactly?
```
$ odo push --debug
Validation
 ✓  Validating the devfile [47060ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [1ms]

Applying URL changes
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs
 ✓  Checking file changes for pushing [493558ns]
 ✓  Syncing files to the component [135ms]

Executing devfile commands for component nodejs
 ✓  Executing install command "npm install", if not running [1s]
 ✓  Executing debug command "npm run debug", if not running [1s]
```

$ odo push 
```
Validation
 ✓  Validating the devfile [22081ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [2ms]

Applying URL changes
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs
 ✓  Checking file changes for pushing [634649ns]
 ✓  No file changes detected, skipping build. Use the '-f' flag to force the build.

Pushing devfile component nodejs
 ✓  Changes successfully pushed to component

```

if you check the logs, debug command is running not the run command
```
$ odo log -f
Debugger listening on ws://127.0.0.1:5858/d6d8cdc8-d412-4f8d-8a29-bf016c22a36d
For help, see: https://nodejs.org/en/docs/inspector
time="2020-08-19T14:10:16Z" level=debug msg="wait program exit" program=debugrun 
```

## Actual behavior
odo push should run the run command not debug command

## Expected behavior
odo push should stop the debug command and start run command, if run after odo push --debug.

### PR DESCRIPTION
**Pull request by adisky**
**What type of PR is this?**
 /kind feature

**What does does this PR do / why we need it**:
Adds Support for `hotReloadCapable` field, it is a replacement of attribute `restart:false`.
This PR also fixes a bug while switching mode from debug to run.

**Which issue(s) this PR fixes**:

Fixes #3548 
Fixes #3784

**PR acceptance criteria**:

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation (would raise separate PR for doc )

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

File: tests/integration/devfile/cmd_devfile_push_test.go [hunk 1]
```diff
@@ -462,8 +463,8 @@ var _ = Describe("odo devfile push command tests", func() {
 			utils.ExecWithInvalidCommandGroup(context, cmpName, namespace)
 		})
 
-		It("should not restart the application if restart is false", func() {
-			utils.ExecWithRestartAttribute(context, cmpName, namespace)
+		It("should not restart the application if it is hot reload capable", func() {
```
Comment (prietyc123) [inline tests/integration/devfile/cmd_devfile_push_test.go:466]: Can we also consider default case. What if we don't set hot reload to `true`. According to https://github.com/devfile/api/pull/96/files#diff-75282a0f7931c780b57f6bf165b6fc9fR274-R277 the description says it should not restart with hot reload enable. However with default case it should. Correct me if I am wrong

**Review by prietyc123 — CHANGES_REQUESTED** [review]

File: pkg/devfile/parser/data/common/types.go [hunk 2]
```diff
@@ -151,6 +151,13 @@ type Exec struct {
 
 	// Working directory where the command should be executed
 	WorkingDir string `json:"workingDir,omitempty" yaml:"workingDir,omitempty"`
+
+	// +optional
+	// Whether the command is capable to reload itself when source code changes.
+	// If set to `true` the command won't be restarted and it is expected to handle file changes on its own.
+	//
+	// Default value is `false`
```
Comment (prietyc123) [inline pkg/devfile/parser/data/common/types.go:159]: Can you please also specify in comment, what would default value do?

**Review by prietyc123 — CHANGES_REQUESTED** [review]

File: pkg/devfile/parser/data/common/types.go [hunk 3]
```diff
@@ -151,6 +151,13 @@ type Exec struct {
...
+	HotReloadCapable bool `json:"hotReloadCapable,omitempty" yaml:"hotReloadCapable,omitempty"`
```
Comment (prietyc123) [inline pkg/devfile/parser/data/common/types.go:160]: IMO we need doc updation also for HotReloadCapable behaviour.

**Review by prietyc123 — CHANGES_REQUESTED** [review]

**Comment by prietyc123** [comment]
@adisky pr needs rebase

File: pkg/devfile/parser/data/common/types.go (hunk 3 above)
Comment (adisky) [reply pkg/devfile/parser/data/common/types.go:160]: Already mentioned in PR desc that i would raise a separate PR for doc

File: pkg/devfile/parser/data/common/types.go (hunk 2 above)
Comment (adisky) [reply pkg/devfile/parser/data/common/types.go:159]: firstly this description directly comes from upstream reference https://github.com/devfile/api/blob/master/schemas/devfile.json#L208

secondly `hotReload` is `bool`, `true` condition is sufficiently explained, false would be opposite of that.

File: tests/integration/devfile/cmd_devfile_push_test.go (hunk 1 above)
Comment (adisky) [reply tests/integration/devfile/cmd_devfile_push_test.go:466]: @prietyc123 sure, will add more tests.

File: pkg/devfile/adapters/common/types.go [hunk 4]
```diff
@@ -43,6 +43,7 @@ type PushParameters struct {
 	EnvSpecificInfo          envinfo.EnvSpecificInfo // EnvSpecificInfo contains infomation of env.yaml file
 	Debug                    bool                    // Runs the component in debug mode
 	DebugPort                int                     // Port used for remote debugging
+	ModeChanged              bool                    // It determines if mode is changed from run to debug or vice versa
```
Comment (mik-dass) [inline pkg/devfile/adapters/common/types.go:46]: I think this should be descriptive like `RunModeChanged`

File: tests/integration/devfile/utils/utils.go [hunk 5]
```diff
@@ -308,14 +308,14 @@ func ExecPushWithNewFileAndDir(projectDirPath, cmpName, namespace, newFilePath,
...
-func ExecWithRestartAttribute(projectDirPath, cmpName, namespace string) {
+// ExecWithHotReload executes odo push with hot reload true
+func ExecWithHotReload(projectDirPath, cmpName, namespace string) {
 	args := []string{"create", "nodejs", cmpName}
 	args = useProjectIfAvailable(args, namespace)
 	helper.CmdShouldPass("odo", args...)
 
 	helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), projectDirPath)
-	helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-with-restart.yaml"), filepath.Join(projectDirPath, "devfile.yaml"))
+	helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-with-hotReload.yaml"), filepath.Join(projectDirPath, "devfile.yaml"))
```
Comment (mik-dass) [inline tests/integration/devfile/utils/utils.go:318]: can we improve this test by checking in the logs? I mean checking in the logs for the absence of a kill string like
```
time="2020-08-20T11:14:05Z" level=error msg="program stopped with error:signal: killed" program=devrun 
time="2020-08-20T11:14:05Z" level=info msg="Stopped by user, don't start it again" program=devrun 
```

File: pkg/devfile/parser/data/common/types.go (hunk 3 above)
Comment (prietyc123) [reply pkg/devfile/parser/data/common/types.go:160]: I didn't see that anywhere in the PR description 👀 

I would say adding the doc with this PR itself makes more sense than in other pr. I really think that to be a part of this issue. TBH its a better practice to get everything included in one PR. So, can you please add it in this PR itself.

**Review by prietyc123 — CHANGES_REQUESTED** [review]
As per the description this PR also resolves #3784 . Can we also include tests to verify that `odo push` works after `odo push --debug`. AFAIK we don't test which verifies #3784 content. @adisky Please add those missing test as well.

File: tests/integration/devfile/cmd_devfile_push_test.go (hunk 1 above)
Comment (adisky) [reply tests/integration/devfile/cmd_devfile_push_test.go:470]: Done.

File: pkg/devfile/parser/data/common/types.go (hunk 3 above)
Comment (adisky) [reply pkg/devfile/parser/data/common/types.go:160]: Added.

File: pkg/devfile/adapters/common/types.go (hunk 4 above)
Comment (adisky) [reply pkg/devfile/adapters/common/types.go]: Done

File: tests/integration/devfile/utils/utils.go (hunk 5 above)
Comment (adisky) [reply tests/integration/devfile/utils/utils.go]: Done.

**Comment by adisky** [comment]
@prietyc123 @mik-dass All your comments are addressed PTAL.

retesting to avoid any potential flake introduction.
/retest

File: pkg/devfile/adapters/common/utils.go [hunk 6]
```diff
@@ -223,20 +232,15 @@ func IsPortPresent(endpoints []common.Endpoint, port int) bool {
...
-	if ok {
-		restart, err = strconv.ParseBool(rs)
-		// Ignoring error here as restart is true for all error and default cases.
-		if err != nil {
-			klog.V(4).Info("Error converting restart attribute to bool")
-		}
-	}
+// GetCommandsMap returns a mapping of all of devfile command names to their corresponding DevfileCommand struct
+// Allowing us to easily retrieve the DevfileCommand of any command listed in a composite command
+func GetCommandsMap(commands []common.DevfileCommand) map[string]common.DevfileCommand {
```
Comment (mik-dass) [inline pkg/devfile/adapters/common/utils.go:237]: This is not used anywhere. I guess it was removed recently.

Comment (adisky) [reply pkg/devfile/adapters/common/utils.go]: ya, thanks it got added during rebase, fixed now

**Comment by prietyc123** [comment]
Hitting the failure on travis CI
```
[odo]  ✗  Failed to start component with name xvhhxb. Error: Failed to create the component: command execution failed: unable to exec command [/opt/odo/bin/supervisord -c /opt/odo/conf/devfile-supervisor.conf -d]: 
[odo] : error while streaming command: error sending request: Post https://127.0.0.1:8443/api/v1/namespaces/eecbamibwn/pods/xvhhxb-64595f4749-99ftb/exec?command=%2Fopt%2Fodo%2Fbin%2Fsupervisord&command=-c&command=%2Fopt%2Fodo%2Fconf%2Fdevfile-supervisor.conf&command=-d&container=runtime&stderr=true&stdout=true: read tcp 127.0.0.1:32930->127.0.0.1:8443: read: connection reset by peer
Deleting project: eecbamibwn
Running odo with args [odo project delete eecbamibwn -f]
[odo]  ✗  The project eecbamibwn does not exist. Please check the list of projects using `odo project list`
• Failure [81.980 seconds]
odo devfile push command tests
/home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_push_test.go:21
  Verify devfile push works
  /home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_push_test.go:79
    Ensure that push -f correctly removes local deleted files from the remote target sync folder [It]
    /home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_push_test.go:572
    No future change is possible.  Bailing out early after 78.827s.
    Running odo with args [odo push --namespace eecbamibwn]
    Expected
        <int>: 1
    to match exit code:
        <int>: 0
```
which has already been tracked in https://github.com/openshift/odo/issues/3731

**Comment by mik-dass** [comment]
```
error: some steps failed:
  * could not run steps: step [release:latest] failed: release "release-latest" failed: pod didn't start running within 30 minutes: 
* Container artifacts is not ready with reason Completed and message 
```
/retest

**Review by mik-dass — APPROVED** [review]
/lgtm

**Pull request merged**